storage/series/lookup/entry.go

```diff
@@ -0,0 +1,248 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package lookup
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/m3db/m3db/storage/index"
+	"github.com/m3db/m3db/storage/series"
+	xtime "github.com/m3db/m3x/time"
+)
+
+const (
+	maxUint64 = ^uint64(0)
+	maxInt64  = int64(maxUint64 >> 1)
+)
+
+// Entry is the entry in the shard ident.ID -> series map. It has additional
+// members to track lifecycle and minimize indexing overhead.
+// NB: users are expected to use `NewEntry` to construct these objects.
+type Entry struct {
+	Series         series.DatabaseSeries
+	Index          uint64
+	curReadWriters int32
+	reverseIndex   entryIndexState
+}
+
+// ensure Entry satisfies the `index.OnIndexSeries` interface.
+var _ index.OnIndexSeries = &Entry{}
+
+// NewEntry returns a new Entry.
+func NewEntry(series series.DatabaseSeries, index uint64) *Entry {
+	entry := &Entry{
+		Series: series,
+		Index:  index,
+	}
+	entry.reverseIndex.states = entry.reverseIndex._staticAloc[:0]
+	return entry
+}
+
+// ReaderWriterCount returns the current ref count on the Entry.
+func (entry *Entry) ReaderWriterCount() int32 {
+	return atomic.LoadInt32(&entry.curReadWriters)
+}
+
+// IncrementReaderWriterCount increments the ref count on the Entry.
+func (entry *Entry) IncrementReaderWriterCount() {
+	atomic.AddInt32(&entry.curReadWriters, 1)
+}
+
+// DecrementReaderWriterCount decrements the ref count on the Entry.
+func (entry *Entry) DecrementReaderWriterCount() {
+	atomic.AddInt32(&entry.curReadWriters, -1)
+}
+
+// IndexedForBlockStart returns a bool to indicate if the Entry has been successfully
+// indexed for the given index blockstart.
+func (entry *Entry) IndexedForBlockStart(indexBlockStart xtime.UnixNano) bool {
+	entry.reverseIndex.RLock()
+	isIndexed := entry.reverseIndex.indexedWithRLock(indexBlockStart)
+	entry.reverseIndex.RUnlock()
+	return isIndexed
+}
+
+// NeedsIndexUpdate returns a bool to indicate if the Entry requires to be indexed
```

Comment:
super nit, but "requires to be" sounds super weird to me. Presumably this comment was already here though

Reply:
It was, but I can update.